### PR TITLE
common,rbd-nbd: fix up prefork behavior vs AsyncMessenger singletons

### PR DESCRIPTION
**Pull request by liewegas**
http://tracker.ceph.com/issues/23143

There seems to be something wrong with the code that shuts down the Stack singleton and starts it up again.  I was having trouble debugging the stop()->start() sequence.  This PR aims to avoid it entirely by destroying and recreating the Stack item, and then eliminating the cases where we have a messenger that survives across a fork.  Mostly this means adding preforker to osd and mds.  rbd-nbd also had semi-broken behavior here that we fix.

This is manifesting now because the wip-config branch (which should merge shortly) creates a messenger during early startup and tears it down again.

### CODE REVIEW

File: src/tools/rbd_nbd/rbd-nbd.cc [hunk 1]
```diff
@@ -855,9 +854,8 @@ static int do_map(int argc, const char *argv[], Config *cfg)
     cout << cfg->devpath << std::endl;
```
Comment (mingshuaiwang) [inline src/tools/rbd_nbd/rbd-nbd.cc]: hi , 
after fork,  function global_init_postfork_start closed stdin and stdout， so this log message will not be printed?

**Comment by yuyuyu101** [comment]
does this async change solve the MonClient get config problem?

**Comment by liewegas** [comment]
Not yet.. I broke something again while I was rebasing and cleaning up.

**Comment by yuyuyu101** [comment]
I see, you move daemonize after bind.

plz refer to https://github.com/ceph/ceph/pull/13384, I met some problem too

**Comment by yuyuyu101** [comment]
https://www.spinics.net/lists/ceph-devel/msg31823.html

**Comment by liewegas** [comment]
http://pulpito.ceph.com/sage-2018-03-05_14:33:23-rbd-wip-sage3-testing-2018-03-04-1532-distro-basic-smithi/

working!

**Comment by liewegas** [comment]
@yuyuyu101 if I follow the linked discussion, they're saying using preforker (and this change in general) is a good thing, right?  Except that we probably would need to make sure the wip-config change avoids using rdma for the short-lived config fetching sequence...

**Comment by yuyuyu101** [comment]
This change is good to me. But  I think we need to run a large qa test to verify. From https://github.com/ceph/ceph/pull/13384 qa result, fork before bind will result to some lock tests failed which rely on msgr init work.

**Comment by liewegas** [comment]
http://pulpito.ceph.com/sage-2018-03-06_13:22:25-rados-wip-sage3-testing-2018-03-05-1355-distro-basic-smithi/
http://pulpito.ceph.com/sage-2018-03-06_19:34:43-rados-wip-sage3-testing-2018-03-05-1355-distro-basic-smithi

http://pulpito.ceph.com/sage-2018-03-06_13:25:43-rbd-wip-sage3-testing-2018-03-05-1355-distro-basic-smithi/
http://pulpito.ceph.com/sage-2018-03-06_19:33:52-rbd-wip-sage3-testing-2018-03-05-1355-distro-basic-smithi

http://pulpito.ceph.com/sage-2018-03-06_13:26:55-fs-wip-sage3-testing-2018-03-05-1355-distro-basic-smithi/
http://pulpito.ceph.com/sage-2018-03-06_19:34:22-fs-wip-sage3-testing-2018-03-05-1355-distro-basic-smithi

the java failure in fs is fixed in #20172

**Pull request merged**